### ISSUE TITLE
Mode Val save_txt&save_conf

### ISSUE THREAD

**Issue by Petros626** [ultralytics/ultralytics#20441]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

When I validate my model and save the detection results (`save_txt`) together prediction scores (`save_conf`), the number of gained labels (3696) doesn't match my original validation labels (3699).

Why 3 labels are missing and how can I save this format `class_id x y w h r(rad) conf` instead of the default one `class_id x1 y1 x2 y2 x3 y3 x4 y4 conf`?

Thank you

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Petros626, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users, where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response 🤖. An Ultralytics engineer will assist you here soon!

**Comment by Y-T-G** [comment]
What batch size did you use for validation?

**Comment by Y-T-G** [comment]
What do you mean the labels are missing? Are you referring to missing txt files or missing instances inside txt files? How did you count the original number?

You can change this for xywhr

https://github.com/ultralytics/ultralytics/blob/e1d742500fbfc24a08c4eb664d0111eaffc8827a/ultralytics/engine/results.py#L717

**Comment by Petros626** [comment]
> What batch size did you use for validation?

16

**Comment by Petros626** [comment]
> What do you mean the labels are missing? Are you referring to missing txt files or missing instances inside txt files? How did you count the original number?
> 
> You can change this for xywhr
> 
> [ultralytics/ultralytics/engine/results.py](https://github.com/ultralytics/ultralytics/blob/e1d742500fbfc24a08c4eb664d0111eaffc8827a/ultralytics/engine/results.py#L717)
> 
> Line 717 in [e1d7425](/ultralytics/ultralytics/commit/e1d742500fbfc24a08c4eb664d0111eaffc8827a)
> 
>  line = (c, *(d.xyxyxyxyn.view(-1) if is_obb else d.xywhn.view(-1)))

I'm referring to missing txt files, got 3699 GT labels in folder `images/val/`, but only 3696 saved predictions for them.

I expected that in this line `is_obb` is True to return the right prediction format?

**Comment by Y-T-G** [comment]
If nothing is detected in an image, no labels are saved for that image.

**Comment by glenn-jocher** [comment]
Thanks for sharing your solution for saving results in the `xywhr` format. This is indeed the correct approach for modifying the output format to `class_id x y w h r(rad) conf`. Y-T-G correctly explained why 3 files were missing - the Ultralytics validation process doesn't create text files for images where no objects were detected. If you need files for every image (even empty ones), you could add a post-processing step to create empty files for the missing image IDs. Your code modification is a helpful reference for others who need rotated bounding box format instead of the default polygon representation.

**Comment by Petros626** [comment]
I'd like to add one further question related to the predictions results.

When I use `conf=0.5` and `iou=0.5` to save my predictions, do I have to set `iou` again, if I want results with another threshold or it is calculated automatically like in the table output?

When a low `iou` increase FP detections, why the metrics rise?

**Comment by Y-T-G** [comment]
`iou` that use provide is for NMS. The IoU used for matching predictions to ground truth can't be specified. It uses a range from 0.5 to 0.95.
